### PR TITLE
KAFKA-3514: Part II, Choose tasks with data on all partitions to process

### PR DESCRIPTION
**Pull request by guozhangwang**
1. In each iteration, decide if a task is processable if all of its partitions contains data, so it can decide which record to process next.

1.a Add one exception that, if the task indeed have data on some but not all of its partitions, we only consider as not processable for some finite round of iterations.

1.b Add a task-level metric to record whenever we are forced to process a task that is only "partially data available", since it may leads to non-determinism.

2. Break the main loop on put-raw-data and process-them. Since now not all data put into the queue would be processed completely within a single iteration.

NOTE that within an iteration, if a task has exhausted one of its queue it will still be processed, since we only update `processable` list once in each iteration, I'm improving on this on the follow-up part III PR.

3. Found and fixed a bug in metrics recording: the taskName and sensorName parameters were exchanged.

4. Optimized task stream time computation again since our current partition stream time reasoning has been simplified.

5. Added unit tests.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by guozhangwang** [comment]
@mjsax @bbejeck Okay part II starts to have some controversial changes. I intentionally left the tests yet to be added, so that you guys can take a look first and let me know wdyt.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java [hunk 1]
```diff
@@ -826,18 +826,21 @@ long runOnce(final long recordsProcessedBeforeCommit) {
             }
         }
 
-        if (records != null && !records.isEmpty() && taskManager.hasActiveRunningTasks()) {
+        if (records != null && !records.isEmpty()) {
             streamsMetrics.pollTimeSensor.record(computeLatency(), timerStartedMs);
             addRecordsToTasks(records);
+        }
+
+        if (taskManager.hasActiveRunningTasks()) {
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:839]: why this change?

**Review by bbejeck — COMMENTED** [review]
Looks reasonable to me, I only have one very minor comment.  

Also while beyond the scope of this PR, would we want to consider making `WAIT_ON_PARTIAL_INPUT` configurable?  

Additionally,  here if just one queue is empty we'll not process until max wait attempts.  Would we want to consider taking into the consideration the percentage of empty queues? Just a thought though.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java (hunk 1 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:839]: As in description :)

```
Break the main loop on put-raw-data and process-them. Since now not all data put into the queue would be processed completely within a single iteration.
```

**Comment by guozhangwang** [comment]
> Also while beyond the scope of this PR, would we want to consider making WAIT_ON_PARTIAL_INPUT configurable?

We can, I'd just want to make sure this is the way we want to go before adding more LOC.

**Comment by guozhangwang** [comment]
Comparing trunk:

https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1865

with this branch:

https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1866

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java [hunk 2]
```diff
@@ -87,9 +88,13 @@ int maybeCommit() {
...
-        final Iterator<Map.Entry<TaskId, StreamTask>> it = running.entrySet().iterator();
-        while (it.hasNext()) {
-            final StreamTask task = it.next().getValue();
+
+        // only iterate over tasks who has data in all its input topic partitions
+        final Map<TaskId, StreamTask> processable = running.entrySet().stream()
+                .filter(entry -> entry.getValue().isProcessable())
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+        for (StreamTask task : processable.values()) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java]: nit: add `final`

**Comment by guozhangwang** [comment]
One thing I've observed while debugging the integration test is that, now we will always NOT drain the queue in an iteration for any tasks that have more than on input topic partition, since once you have drained one of partition's queue you would stop processing the others until next iteration when new data is fetched. This is indeed our wanted behavior, but I do not know if it will have any impact on performance, hence triggered the above tests.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java [hunk 3]
```diff
@@ -108,7 +112,7 @@
 
             // add the operation metrics with additional tags
             final Map<String, String> tagMap = metrics.tagMap("task-id", taskName);
-            taskCommitTimeSensor = metrics.taskLevelSensor("commit", taskName, Sensor.RecordingLevel.DEBUG, parent);
+            taskCommitTimeSensor = metrics.taskLevelSensor(taskName, "commit", Sensor.RecordingLevel.DEBUG, parent);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:115]: Can you open PRs for older branches to back-port this fix?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java [hunk 4]
```diff
@@ -263,6 +279,30 @@ public void resume() {
         log.debug("Resuming");
     }
 
+    /**
+     * An active task is processable if its buffer contains data for all of its input source topic partitions
+     */
+    public boolean isProcessable() {
+        final Map<TopicPartition, RecordQueue> partitionQueues = partitionGroup.partitionQueues();
+
+        for (RecordQueue queue : partitionQueues.values()) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: nit: add `final`

File: streams/src/main/java/org/apache/kafka/streams/state/internals/NamedCache.java [hunk 5]
```diff
@@ -370,7 +370,7 @@ private NamedCacheMetrics(final StreamsMetricsImpl metrics, final String cacheNa
                 "record-cache-id", "all",
                 "task-id", taskName
             );
-            final Sensor taskLevelHitRatioSensor = metrics.taskLevelSensor("hitRatio", taskName, Sensor.RecordingLevel.DEBUG);
+            final Sensor taskLevelHitRatioSensor = metrics.taskLevelSensor(taskName, "hitRatio", Sensor.RecordingLevel.DEBUG);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/NamedCache.java:373]: As above: can we do a new PR to back port this fix to older branches?

File: streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/StreamToTableJoinScalaIntegrationTestBase.scala [hunk 6]
```diff
@@ -129,9 +125,9 @@ class StreamToTableJoinScalaIntegrationTestBase extends JUnitSuite with StreamTo
       // consume and verify result
       val consumerConfig = getConsumerConfig()
 
-      IntegrationTestUtils.waitUntilMinKeyValueRecordsReceived(consumerConfig,
-                                                               outputTopic,
-                                                               expectedClicksPerRegion.size)
+      IntegrationTestUtils.waitUntilExactKeyValueRecordsReceived(consumerConfig,
```
Comment (mjsax) [inline streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/StreamToTableJoinScalaIntegrationTestBase.scala]: It's unclear why we need "exactly" now?

**Review by mjsax — COMMENTED** [review]
I agree with @bbejeck comments about making "blocking" behavior configurable with a follow up KIP. For ease of configuration, I think a `max.block.task.ms` config (or similar name) would make more sense than a number of retries as this PR introduces.

Meta comment: on startup, should we have a "bootstrap" phase in which we try to fetch data for all tasks? Ie, a loop that polls until we got data for all tasks (ie, all tasks are processable), or poll() does not return any new data (ie, we would pause partitions for which we have data already).

Also, should we pause() partitions for which tasks still buffer data to make `poll()` more effective?

**Comment by guozhangwang** [comment]
retest this please

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java (hunk 3 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:115]: yup.

File: streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/StreamToTableJoinScalaIntegrationTestBase.scala (hunk 6 above)
Comment (guozhangwang) [reply streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/StreamToTableJoinScalaIntegrationTestBase.scala]: Because the integration test used to expect exactly three records only in `waitUntil`MinKeyValue` but more could be more output, and we should actually check that for each key, the final result is correct. Or we can do the same as I did in Java integration test, which is sending one record from producer to input topics at a time, and then immediately check for the intermediate result from the output topic, and then send one record again. I felt this may be simpler to check without loss of generality.

With this case because we do not drain in the same iteration, its flakiness gets exposed.

**Comment by guozhangwang** [comment]
> Meta comment: on startup, should we have a "bootstrap" phase in which we try to fetch data for all tasks? Ie, a loop that polls until we got data for all tasks (ie, all tasks are processable), or poll() does not return any new data (ie, we would pause partitions for which we have data already).

For partitions that have high traffic variance, doing this bootstrap may postpone the processing as well, right? What's the motivation that we'd like to pay more latency waiting for all data to be available at the beginning of the processing, than during the middle of the processing?

> 
> Also, should we pause() partitions for which tasks still buffer data to make poll() more effective?

That's a good question, in fact with the current improvement on consumers for round-robin-starvation-avoidance fetch request, maybe pausing / resuming is not really necessary any more, would double check on the consumer code.

**Comment by mjsax** [comment]
About the bootstrapping: I would not wait infinitely. My concerns if more for use cases with large number of partitions. Assume all partitions do have data. A single poll() will not return data for all partitions. Thus, I could imagine, that for larger number of input topic partitions, we might get partial processing because the exhaust the retries early.

In general, it seems to be a good strategy to me, to unblock for partial processing, if a previous poll() did not return data for those partitions. As we know, that the consumer puts partitions that did return data to the end of its internal round robin list, we should be able to reason about this without pausing/resuming of partitions. Details need to be figured out still.

**Comment by guozhangwang** [comment]
Comparing again trunk:

https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1868

with this branch:

https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1869

**Comment by guozhangwang** [comment]
Re-build with a bit optimization: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1870

**Comment by guozhangwang** [comment]
I have confirmed that the performance of this branch

http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2018-07-25--001.1532560979--guozhangwang--K3514-choose-task-with-available-data-on-all-partitions--e021873/report.html

Is comparable to trunk:

http://confluent-kafka-system-test-results.s3-us-west-2.amazonaws.com/2018-07-20--001.1532125658--confluentinc--trunk--c5a4d5c/report.html

**Comment by guozhangwang** [comment]
Note that I've intentionally leave out the pause / resume into another PR (after https://github.com/apache/kafka/pull/5428) since I observed some performance degradation that needs some further investigation, and hence I want to do this in steps and make sure each step does not introduce performance regression. cc @mjsax

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java [hunk 7]
```diff
@@ -263,6 +279,21 @@ public void resume() {
...
+     * An active task is processable if its buffer contains data for all of its input source topic partitions
+     */
+    public boolean isProcessable() {
+        if (partitionGroup.allPartitionsBuffered()) {
+            return true;
+        } else if (partitionGroup.numBuffered() > 0 && --waits < 0) {
+            taskMetrics.taskEnforcedProcessSensor.record();
+            waits = WAIT_ON_PARTIAL_INPUT;
+            return true;
+        } else {
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: nit: we don't need the last branch if both of the previous conditions evaluate to `false` just have a ` return false`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/PartitionGroup.java [hunk 8]
```diff
@@ -165,6 +161,10 @@ int numBuffered() {
         return totalBuffered;
     }
 
+    boolean allPartitionsBuffered() {
+        return allBuffered;
+    }
+
     public void close() {
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/PartitionGroup.java:168]: Although this is pre-existing, on `close` do we want to clear `nonEmptyQueuesByTime` as well?

**Review by bbejeck — APPROVED** [review]
Thanks, @guozhangwang just a couple of minor nits otherwise LGTM.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java [hunk 9]
```diff
@@ -87,9 +88,13 @@ int maybeCommit() {
      */
     int process() {
         int processed = 0;
-        final Iterator<Map.Entry<TaskId, StreamTask>> it = running.entrySet().iterator();
-        while (it.hasNext()) {
-            final StreamTask task = it.next().getValue();
+
+        // only iterate over tasks who has data in all its input topic partitions
+        final Map<TaskId, StreamTask> processable = running.entrySet().stream()
+                .filter(entry -> entry.getValue().isProcessable())
```
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java]: I might be old-fashioned, but this actually seems more roundabout than just putting `if (!task.isProcessable()) continue;` at the beginning of the old while loop body.

Plus, I'm not sure of the performance impact of creating the stream, the filter class, the collector, and the new map on each invocation.

(I feel like I should add the caveat that I prefer FP in general, it just seems a bit heavy in this case to just filter the map before iterating over it anyway)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/PartitionGroup.java [hunk 10]
```diff
@@ -53,12 +56,13 @@ RecordQueue queue() {
...
     }
 
-
     PartitionGroup(final Map<TopicPartition, RecordQueue> partitionQueues) {
         nonEmptyQueuesByTime = new PriorityQueue<>(partitionQueues.size(), Comparator.comparingLong(RecordQueue::timestamp));
         this.partitionQueues = partitionQueues;
         totalBuffered = 0;
+        allBuffered = false;
         streamTime = RecordQueue.NOT_KNOWN;
+        numQueues = this.partitionQueues.keySet().size();
```
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/PartitionGroup.java]: this could just be `this.partitionQueues.size()`.

Is it really that valuable to cache this number? I'm sure it's already cached on most (if not all) Map implementations.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/PartitionGroup.java [hunk 11]
```diff
@@ -81,15 +85,14 @@ StampedRecord nextRecord(final RecordInfo info) {
 
                 if (!queue.isEmpty()) {
                     nonEmptyQueuesByTime.offer(queue);
+                } else {
+                    // if a certain queue has been drained, reset the flag
+                    allBuffered = false;
```
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/PartitionGroup.java]: Since we have both branches here, maybe we can invert the condition to avoid unnecessary negation (just to improve readability)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/PartitionGroup.java [hunk 12]
```diff
@@ -106,17 +109,22 @@ int addRawRecords(final TopicPartition partition, final Iterable<ConsumerRecord<
...
-            computeStreamTime();
+            // if all partitions now are non-empty, set the flag and compute the stream time
+            if (nonEmptyQueuesByTime.size() == numQueues) {
+                allBuffered = true;
+
+                // since we may enforce processing even if some queue is empty, it is possible that after some
+                // raw data has been added to that queue the new partition's timestamp is even smaller than the current
+                // stream time, in this case we should not update.
+                final long newTimestamp = nonEmptyQueuesByTime.peek().timestamp();
+                streamTime = newTimestamp > streamTime ? newTimestamp : streamTime;
```
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/PartitionGroup.java]: nit: is `streamTime = Math.max(streamTime, nonEmptyQueuesByTime.peek().timestamp());` nicer?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java (hunk 3 above)
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:115]: ah geez; good catch.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java (hunk 7 above)
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java]: :) FWIW, I have the opposite opinion. It seems clearer to me to have all the branches return (and all the returns in branches) rather than have conditionals that might return followed by a return in a separate statement.

That said, it's clearly equivalent either way...

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/PartitionGroupTest.java [hunk 13]
```diff
@@ -108,7 +108,7 @@ public void testTimeTracking() {
         assertEquals(5, group.numBuffered());
         assertEquals(2, group.numBuffered(partition1));
         assertEquals(3, group.numBuffered(partition2));
-        assertEquals(2L, group.timestamp());
+        assertEquals(1L, group.timestamp());
```
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/PartitionGroupTest.java:111]: Interesting. So the stream time is `max(oldStreamTime, currentRecordTime)`, while we pick the current record as the `record with min timestamp among queue heads`.

This is actually different than the prior stream time definition, which would be computed from the queues regardless of the current record's timestamp.

I like the new definition better, as the observer can determine after the fact what the stream time was at any point knowing only the records that got processed (not needing to know what was queued). It also means you know stream time can only advance when you get the next record, not when you add new records.


---

Actually upon re-reading, enqueuing can change the stream time only when the group transitions from missing data from a partition to having data for all partitions. Is it actually necessary for stream time to update in such a scenario? Can we simplify both the definition and implementation by sticking with the definition I wrote down?

If not, maybe we should update this scenario to demonstrate that transition.

Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/PartitionGroupTest.java:111]: Can you also update or remove the explanatory comments? (Update if the behavior might be confusing still, remove if it's obvious under these much simpler semantics)

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/PartitionGroupTest.java [hunk 14]
```diff
@@ -206,7 +206,7 @@ public void testTimeTracking() {
         assertEquals(0, group.numBuffered());
         assertEquals(0, group.numBuffered(partition1));
         assertEquals(0, group.numBuffered(partition2));
-        assertEquals(5L, group.timestamp());
+        assertEquals(6L, group.timestamp());
```
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/PartitionGroupTest.java:209]: I'm not opposed to leaving the whole scenario in place, but fwiw it could be simpler now, since enqueuing no longer updates the stream time.

**Review by vvcephei — COMMENTED** [review]
This is awesome. I have one high-level comment in PartitionGroupTest about potentially simplifying the streamTime definition further.

I also wonder if we should explicitly write down the definition of streamTime in a comment in PartitionGroup.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/PartitionGroup.java (hunk 8 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/PartitionGroup.java:168]: @bbejeck We call `clear()` in `closeTopology` and `close()` in `closeSuspended`, i.e. the `nonEmptyQueuesByTime` should already be cleared when `close()` is called. But I agree with you that to to be on the safer side we can always just call clear() inside close(). Does that sound good to you?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java (hunk 9 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java]: The key is that we want to call `task.IsProcessable()` only once within each StreamThread iteration, since calling it N times will cause it to be forced to process. On the other hand, I'm not sure of the performance impact of this line either, so if there are any better ideas on achieving the following:

1. We only check `task.IsProcessable()` once within the iteration, after polled records have been put into the queue.

2. During processing, if a task is no longer "buffered on all partitions", we stop iterating it but still looping over others.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/PartitionGroup.java (hunk 10 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/PartitionGroup.java]: Ack, will remove this.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/PartitionGroup.java (hunk 11 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/PartitionGroup.java]: ack.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/PartitionGroup.java (hunk 12 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/PartitionGroup.java]: ack.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/PartitionGroupTest.java (hunk 13 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/PartitionGroupTest.java:111]: Sure, I will update the comments in PartitionGroup.

About whether we should still update stream time when filling in records to an empty partition, my gut feeling is yes. Consider the following example of two partitions in a task:

P1 (0): [0,1,2,3,4]
P2 (100): [100,101,102,103,104]

Again the digits above are record timestamps. When processing this task we will pop out records from P1 until it is empty. When a partition becomes empty, its stream time stays as the timestmap of the last popped record, hence P1's st (stream time) stays as 4, and hence the task st stays as 4.

P1 (4): []
P2 (100): [100,101,102,103,104]

Now consider we polled more records into P:

P1 (5): [5,6,7,8]
P2 (100): [100,101,102,103,104]

P1's st will be updated immediately as 5. If we do not update immediately, the task st is still 4, and only when record with timestamp 5 has been popped and processed, the task st will be "jumpping" from 4 to 6.

Of course, one can argue to favor "let the st jump in this case" over "let the st increase after adding new data, and hence users may observe st being increased without processing even a record yet". Personally I'm in favor of the latter, but I'm open to be convinced :)

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/PartitionGroupTest.java (hunk 14 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/PartitionGroupTest.java:209]: Enqueuing may still update the st right now (see my comment above).

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/PartitionGroupTest.java (hunk 13 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/PartitionGroupTest.java:111]: Another note: the above example is that we do not yet enforce processing before new records are enqueued to P1, otherwise it will look like

P1 (4): []
P2 (104): []

I.e. users processing records with ts 100+ while the stream time stays at 4. This is intentional in my design (users would be aware that some enforced processing is happening as st did not increment) but I'm also open for questions / arguments against it.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java (hunk 9 above)
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java]: I'm not sure I follow; putting `(!task.isProcessable()) continue;` in between lines 97 and 98 should be exactly the same as filtering this map before the loop.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/PartitionGroupTest.java (hunk 13 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/PartitionGroupTest.java:111]: It seems like it would be sensible either way.

I think the distinction really comes down to situations in which you observe the stream time *without* processing a record. I think that we only look at the stream time while we are processing a record, so there's actually no need to update the stream time during enqueue.

Actually, I didn't completely follow the "jumping" behavior in your example.

Starting from 
P1 (4): []
P2 (100): [100,101,102,103,104]
ST:4

we poll more records into P1:
P1 (5): [5,6,7,8]
P2 (100): [100,101,102,103,104]
ST:4

Then when we grab a record (5) to process, ST would advance to 5, no?

P1 (6): [6,7,8]
P2 (100): [100,101,102,103,104]
ST:5

In other words, a processor would observe:
record time = 0; stream time = 0
record time = 1; stream time = 1
record time = 2; stream time = 2
record time = 3; stream time = 3
record time = 4; stream time = 4
(processing pauses for a while while we wait for more records on P1)
record time = 5; stream time = 5
... 6, 7, ...
record time = 8; stream time = 8
(processing pauses for a while, until we give up on P1)
record time = 100; stream time = 8
record time = 101; stream time = 8
... etc (stream time doesn't advance any more until we see another record on P1)

I believe this sequence of observations is the same whether you advance stream time during enqueue or not.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java (hunk 9 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java]: Ah, I'm confusing myself with part II and part III PRs :) Yes in this PR it is the same, but in https://github.com/apache/kafka/pull/5428 I did some further modifications so that `isProcessable` is only called once per iteration. The reason I did not put that into this PR is only because I am still investigating some unexpected perf diffs in that PR, and for this PR at least I can confirm that no perf diffs were introduced.

But since we have talked about it, I may just move that logic from 5428 to this one and see if it stays the same.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/PartitionGroupTest.java (hunk 13 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/PartitionGroupTest.java:111]: You are right, I was confused before with the case

P1 (6): [6,7,8]
P2 (100): [100,101,102,103,104]
Processing 5, ST:5

And thought it was

P1 (6): [6,7,8]
P2 (100): [100,101,102,103,104]
Processing 5, ST:6

Since the P1's ts is already 6. But we are not setting task st as the min of all partition's ts anymore, but only the the popped record's ts if it is larger. 

Going back to your original comment, today we may still look at the task st before even processing a record after enqueuing records, think: ST based punctuation. But since we only consider updating ST if `allBuffered = true;` anyways, it means we will definitely tries to process some data right away within in iteration, so it is indeed not necessary to update the ST here.

Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/PartitionGroupTest.java:111]: Got it. Thanks for the clarification!

**Comment by guozhangwang** [comment]
The newest PR has a performance boost from 25% to 2.5X compared with trunk on various simple benchmarks:

http://confluent-kafka-system-test-results.s3-us-west-2.amazonaws.com/2018-07-31--001.1533078590--confluentinc--trunk--495c78d/report.html

This branch:

http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2018-08-02--001.1533173478--guozhangwang--K3514-choose-task-with-available-data-on-all-partitions--bce7fc6/report.html

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java [hunk 15]
```diff
@@ -89,14 +88,12 @@ int maybeCommit() {
     int process() {
         int processed = 0;
 
-        // only iterate over tasks who has data in all its input topic partitions
-        final Map<TaskId, StreamTask> processable = running.entrySet().stream()
-                .filter(entry -> entry.getValue().isProcessable())
-                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+        final Iterator<Map.Entry<TaskId, StreamTask>> it = running.entrySet().iterator();
+        while (it.hasNext()) {
+            final StreamTask task = it.next().getValue();
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java]: Originally I was ok with the Java 8 approach, but this it bit better in that we only traverse tasks once.

**Review by bbejeck — APPROVED** [review]
LGTM with latest updates

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java (hunk 3 above)
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:115]: FWIW, it wouldn't break anything. This is only the name of the sensor, which just needs to be unique, and it's just as unique in this order, unless some other task is named "commit" (which wouldn't happen).

Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:115]: To be clear, we should fix it, just not sure if we need to bother backporting.

**Review by vvcephei — APPROVED** [review]
LGTM! Thanks.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java (hunk 3 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:115]: Cannot follow the argument? It seems to be an easy fix and we should back port all fixes by default and only not back port if there is a good reason for not doing it.

@guozhangwang Is there already a PR or ticket so this is not dropped on the floor?

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:115]: @vvcephei I think it did break sth.. the sensorName is used as part of `fullSensorName = key + "." + sensorName;`. With `sensorName == taskName`, we will not add them since `Sensor s = getSensor(name);` already exists, i.e. we will lose some metrics, right?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/NamedCache.java (hunk 5 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/NamedCache.java:373]: Just a reminder that this should be part of the backport fix.

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java [hunk 16]
```diff
@@ -350,6 +350,33 @@ public static void waitForCompletion(final KafkaStreams streams,
...
+                                                                                    final long waitTime) throws InterruptedException {
+        final List<KeyValue<K, V>> accumData = new ArrayList<>();
+        try (final Consumer<K, V> consumer = createConsumer(consumerConfig)) {
+            final TestCondition valuesRead = new TestCondition() {
+                @Override
+                public boolean conditionMet() {
+                    final List<KeyValue<K, V>> readData =
+                            readKeyValues(topic, consumer, waitTime, expectedRecords.size());
+                    accumData.addAll(readData);
+                    return accumData.containsAll(expectedRecords);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java]: This is not a check for "exactly": if `accumData` contains more record than `expectedRecords` this test would pass. Shouldn't we use `equals()` ?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java [hunk 17]
```diff
@@ -307,97 +308,6 @@ public void testPauseResume() {
         assertEquals(0, consumer.paused().size());
     }
 
-    @SuppressWarnings("unchecked")
-    @Test
-    public void testMaybePunctuateStreamTime() {
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:312]: Why is this removed but not updated? Is is redundant?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java [hunk 18]
```diff
@@ -518,6 +437,61 @@ public void testCancelPunctuateStreamTime() {
...
+    public void shouldBeProcessableIfWaitedForTooLong() {
+        task = createStatelessTask(createConfig(false));
+        task.initializeStateStores();
+        task.initializeTopology();
+
+        assertFalse(task.isProcessable());
+
+        final byte[] bytes = ByteBuffer.allocate(4).putInt(1).array();
+
+        task.addRecords(partition1, Collections.singleton(new ConsumerRecord<>(topic1, 1, 0, bytes, bytes)));
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:483]: We do reset the `wait-counter` to 5 when we add records on purpose?

**Comment by guozhangwang** [comment]
> @guozhangwang Is there already a PR or ticket so this is not dropped on the floor?

I plan to do the cherry-pick right after the PR is merged, so I did not create another ticket for it (am working on the cherry-pick PR now).

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java (hunk 18 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:483]: Yes, we want to only force process once every these times.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java (hunk 17 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:312]: It is renamed to another test function :)

File: streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java (hunk 16 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java]: Good catch, actually we should check that for each key, the expected record is the last record, I'll try to update the function.

**Review by mjsax — APPROVED** [review]
LGMT. Feel free to merge after Jenkins passed.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java (hunk 3 above)
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:115]: @mjsax fair enough!

@guozhangwang I think you're right. The full sensor name is still unique, but it would mess up the ability to unload sensors (we basically wouldn't be able to unload any task level sensors). My bad.

Comment (tedyu) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:115]: Maybe create (internal) SensorName type (alias for String) so that there is no chance of mixup in the future.
